### ISSUE TITLE
'@androidndk//:toolchain-libcpp' does not contain a toolchain for cpu 'darwin'

### ISSUE THREAD

**Issue by UebelAndre** [bazelbuild/rules_android#27]
Hi, I hope this is the right place for this. I'm getting the following build error when I try to build an `android_binary` with an `android_library` that's trying to link a `rust_library`
```
INFO: Invocation ID: 0757cbc6-cf68-4e12-8afe-960a87bfcb01
ERROR: /private/var/tmp/_bazel_user/97e9600557997318199814824b024c95/external/androidndk/BUILD.bazel:41:19: in cc_toolchain_suite rule @androidndk//:toolchain-libcpp: cc_toolchain_suite '@androidndk//:toolchain-libcpp' does not contain a toolchain for cpu 'darwin'
WARNING: errors encountered while analyzing target '//:app': it will not be built
INFO: Analysis succeeded for only 2 of 3 top-level targets
INFO: Analyzed 3 targets (0 packages loaded, 0 targets configured).
INFO: Found 2 targets...
ERROR: command succeeded, but not all targets were analyzed
INFO: Elapsed time: 3.862s, Critical Path: 0.00s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```
I'm trying to write a "Hello, World" app where the string "Hello, World" comes form a Rust library. Is there something I'm doing wrong here?

## Repro Workspace
The zip below contains a repro case. I've also added my `BAZEL.build` and `WORKSPACE` files here for convenience.
[repro.zip](https://github.com/bazelbuild/rules_android/files/5005270/repro.zip)

### `BUILD.bazel`
```
load("@io_bazel_rules_rust//rust:rust.bzl", "rust_library")
load("@build_bazel_rules_android//android:rules.bzl", "android_binary", "android_library")

package(default_visibility = ["//visibility:public"])

rust_library(
    name = "repro",
    srcs = glob(include = ["src/**/*.rs"]),
    crate_type = "cdylib",
    deps = [
        "//cargo:ferris_says",
    ],
    tags = ["manual"],
)

android_library(
    name = "lib",
    srcs = glob(include = ["android/src/main/**/*.java"]),
    custom_package = "com.repro.example",
    manifest = "android/src/main/LibraryManifest.xml",
    resource_files = glob(["android/src/main/res/**/*"]),
    deps = [
        "//:repro",
        "@maven//:androidx_appcompat_appcompat",
        "@maven//:androidx_constraintlayout_constraintlayout",
        "@maven//:androidx_navigation_navigation_fragment",
        "@maven//:androidx_navigation_navigation_ui",
        "@maven//:com_google_android_material_material",
    ],
    tags = ["manual"],
)

android_binary(
    name = "app",
    custom_package = "com.repro.example",
    manifest = "src/main/AndroidManifest.xml",
    manifest_values = {
        "minSdkVersion": "28",
        "targetSdkVersion": "28",
    },
    deps = [":lib"],
)
```

### `WORKSPACE`
```
workspace(name = "repro")

load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")

git_repository(
    name = "bazel_skylib",
    commit = "d35e8d7bc6ad7a3a53e9a1d2ec8d3a904cc54ff7",
    remote = "https://github.com/bazelbuild/bazel-skylib.git",
    shallow_since = "1593183852 +0200",
)

load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
bazel_skylib_workspace()

git_repository(
    name = "io_bazel_rules_rust",
    commit = "f41a008d3e36e70b9deba9e4ed6cd914e8f36a49",
    remote = "https://github.com/bazelbuild/rules_rust.git",
    shallow_since = "1593798220 +0200",
)

load("@io_bazel_rules_rust//:workspace.bzl", "bazel_version")
bazel_version(name = "bazel_version")
load("@io_bazel_rules_rust//rust:repositories.bzl", "rust_repositories")
rust_repositories(version = "1.45.0", edition = "2018")
load("//cargo:crates.bzl", "raze_fetch_remote_crates")
raze_fetch_remote_crates()

git_repository(
    name = "build_bazel_rules_android",
    remote = "https://github.com/bazelbuild/rules_android.git",
    commit = "e8fbc49f913101e846235b9c9a31b3aa9788364a",
    shallow_since = "1593973351 -0700",
)
git_repository(
    name = "rules_jvm_external",
    remote = "https://github.com/bazelbuild/rules_jvm_external.git",
    commit = "a8300d874fcda9b3849c288f4ad57c53a454953e",
    shallow_since = "1594195655 +0800",
)

android_sdk_repository(
    name = "androidsdk",
    api_level = 28,
    build_tools_version = "28.0.3",
)
android_ndk_repository(
    name = "androidndk",
    api_level = 28,
)

load("@rules_jvm_external//:defs.bzl", "maven_install")
maven_install(
    artifacts = [
        "androidx.appcompat:appcompat:1.1.0",
        "com.google.android.material:material:1.0.0",
        "androidx.constraintlayout:constraintlayout:1.1.3",
        "androidx.navigation:navigation-fragment:2.1.0",
        "androidx.navigation:navigation-ui:2.1.0",
    ],
    repositories = [
        "https://jcenter.bintray.com/",
        "https://maven.google.com",
    ],
)
```

**Comment by UebelAndre** [comment]
@timpeut Do you know if this is the correct place to ask this question?

**Comment by katre** [comment]
How are you invoking the build? What's your bazel command line?

This is probably an incompatibility between Android rules (which are not currently using toolchain resolution) and the Rust rules (which do use toolchain resolution). We are working on the former (see https://github.com/bazelbuild/bazel/issues/11749). There should be a way to have your build work currently, by adding the extra flags needed, but I will need a little more information to help figure it out.

**Comment by UebelAndre** [comment]
@katre Hi! Thank you so much for the reply!

I've put together a small workspace to reproduce the issue. Simply running `bazel build //...` should generate the error above.
[repro.zip](https://github.com/bazelbuild/rules_android/files/5005264/repro.zip)

**Comment by katre** [comment]
Thanks for the test case. This has existed since at least Bazel 3.2.0, and possibly earlier.

The problem is that the Android binary is changing the configuration to use the c compiler from the NDK (which makes sense), but the rust rules are trying to access that using the legacy system and an invalid `--cpu` flag (with the deafult value, k8).

I'm not entirely sure where this is happening: it could be rules_rust, rules_android, or Bazel itself. We're in the process of migrating the Android rules to use toolchain resolution, which will also unblock using toolchain resolution for C++ code: this should make this much simpler because each set of rules will be using the same configuration settings to tell what platform to build for.

I'm going to keep poking and see if there's an easy way to get your build working without waiting for those efforts to finish. You might also consider asking on the mailing lists if anyone else is using an Android binary with another set of rules to build binary code: rules_rust, rules_go, and several others are all very similar and a fix for one might work for you, even if it's not specifically for rust.

**Comment by UebelAndre** [comment]
**I have not verified this is resolved**. But I'm gonna close this out hoping it was fixed. If someone is still running into it, please open a new issue and link this one.
